timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java

```diff
@@ -79,11 +102,99 @@
+     *  response may be off by 1 per client that exceeded the limit.
+     */
+    @Test
+    public void exceedingThreadCountLimitsReturns429() {
+        List<RemoteLockService> lockService = ImmutableList.of(getLockService(CLIENT_1));
+
+        assertThat(lockAndUnlockAndCountExceptions(lockService, LOCAL_TC_LIMIT + GLOBAL_TC_LIMIT + FORTY_TWO))
+                .isBetween(FORTY_TWO - lockService.size(), FORTY_TWO);
+    }
+
+    @Test
+    public void multipleClientsCanShareGlobalThreads() {
+        List<RemoteLockService> lockServiceList = ImmutableList.of(
+                getLockService(CLIENT_1), getLockService(CLIENT_2), getLockService(CLIENT_3));
+
+        assertThat(lockAndUnlockAndCountExceptions(lockServiceList, LOCAL_TC_LIMIT + GLOBAL_TC_LIMIT / 3))
+                .isEqualTo(0);
+    }
+
+    @Test
+    public void globalThreadCountLimitAppliesToAllClients() {
+        List<RemoteLockService> lockServiceList = ImmutableList.of(
+                getLockService(CLIENT_1), getLockService(CLIENT_2), getLockService(CLIENT_3));
+
+        assertThat(lockAndUnlockAndCountExceptions(lockServiceList, LOCAL_TC_LIMIT + GLOBAL_TC_LIMIT))
+                .isBetween(2 * GLOBAL_TC_LIMIT - lockServiceList.size(), 2 * GLOBAL_TC_LIMIT);
+    }
+
+    @Test
+    public void clientsCanUseTheirAllowanceWhenGlobalLimitIsReached() throws Exception {
+        RemoteLockService lockService1 = getLockService(CLIENT_1);
+        RemoteLockService lockService2 = getLockService(CLIENT_2);
+        ExecutorService executorService =
+                Executors.newFixedThreadPool(GLOBAL_TC_LIMIT + 2 * LOCAL_TC_LIMIT + FORTY_TWO);
+        List<Future<LockRefreshToken>> initialRequests;
+        List<Future<LockRefreshToken>> secondClientRequests;
+        List<Future<LockRefreshToken>> followUpRequests;
+
+        // This will exhaust all the globally available threads
+        initialRequests = requestLocks(lockService1, GLOBAL_TC_LIMIT + LOCAL_TC_LIMIT + FORTY_TWO, executorService);
+
+        // The following are assigned to CLIENT_2's threads
+        secondClientRequests = requestLocks(lockService2, LOCAL_TC_LIMIT, executorService);
+
+        // CLIENT_1 still cannot acquire a thread
+        followUpRequests = requestLocks(lockService1, FORTY_TWO, executorService);
+
+        assertThat(unlockAndCountExceptions(lockService2, secondClientRequests)).isEqualTo(0);
+        secondClientRequests = requestLocks(lockService2, LOCAL_TC_LIMIT, executorService);
+        assertThat(unlockAndCountExceptions(lockService2, secondClientRequests)).isEqualTo(0);
+
+        // Verify that all the global threads were occupied by CLIENT_1's requests
+        assertThat(unlockAndCountExceptions(lockService1, initialRequests)).isBetween(FORTY_TWO - 1, FORTY_TWO);
+        assertThat(unlockAndCountExceptions(lockService1, followUpRequests)).isBetween(FORTY_TWO - 1, FORTY_TWO);
+    }
+
```

Comment:
Wow, these are some pretty intricate tests. I've had a look and it _seems_ right, will have a closer look again before we merge